Parse/src/main/java/com/parse/ParsePolygon.java

```diff
@@ -0,0 +1,228 @@
+  }
+
+  /**
+   * Creates a new point instance from a {@link Parcel} using the given {@link ParseParcelDecoder}.
+   * The decoder is currently unused, but it might be in the future, plus this is the pattern we
+   * are using in parcelable classes.
+   *
+   * @param source the parcel
+   * @param decoder the decoder
+   */
+  ParsePolygon(Parcel source, ParseParcelDecoder decoder) {
+    setCoordinates(source.readArrayList(null));
+  }
+
+
+  /**
+   * Set coordinates. Valid are Array of GeoPoint, ParseGeoPoint or Location
+   * at least 3 points
+   *
+   * @param coords
+   *          The polygon's coordinates.
+   */
+  public void setCoordinates(List<ParseGeoPoint> coords) {
+    this.coordinates = ParsePolygon._validate(coords);
+  }
+
+  /**
+   * Get coordinates.
+   */
+  public List<ParseGeoPoint> getCoordinates() {
+    return coordinates;
+  }
+
+  /**
+   * Get converts coordinate to JSONArray.
+   */
+  public JSONArray coordinatesToJSONArray() throws JSONException{
+    JSONArray points = new JSONArray();
+    for (ParseGeoPoint coordinate : coordinates) {
+      JSONArray point = new JSONArray();
+      point.put(coordinate.getLatitude());
+      point.put(coordinate.getLongitude());
+      points.put(point);
+    }
+    return points;
+  }
+
+  /**
+   * Checks if this {@code ParsePolygon}; contains {@link ParseGeoPoint}.
+   */
+  public boolean containsPoint(ParseGeoPoint point) {
+    double minX = coordinates.get(0).getLatitude();
+    double maxX = coordinates.get(0).getLatitude();
+    double minY = coordinates.get(0).getLongitude();
+    double maxY = coordinates.get(0).getLongitude();
+
+    for ( int i = 1; i < coordinates.size(); i += 1) {
+        ParseGeoPoint geoPoint = coordinates.get(i);
+        minX = Math.min(geoPoint.getLatitude(), minX);
+        maxX = Math.max(geoPoint.getLatitude(), maxX);
+        minY = Math.min(geoPoint.getLongitude(), minY);
+        maxY = Math.max(geoPoint.getLongitude(), maxY);
+    }
+
+    boolean outside = point.getLatitude() < minX || point.getLatitude() > maxX || point.getLongitude() < minY || point.getLongitude() > maxY;
+    if (outside) {
+        return false;
+    }
+
+    boolean inside = false;
+    for (int i = 0, j = coordinates.size() - 1 ; i < coordinates.size(); j = i++) {
+      double startX = coordinates.get(i).getLatitude();
+      double startY = coordinates.get(i).getLongitude();
+      double endX = coordinates.get(j).getLatitude();
+      double endY = coordinates.get(j).getLongitude();
+
+      boolean intersect = (( startY > point.getLongitude() ) != ( endY > point.getLongitude() ) &&
+          point.getLatitude() < ( endX - startX ) * ( point.getLongitude() - startY ) / ( endY - startY ) + startX);
+
+        if (intersect) {
+            inside = !inside;
+        }
+    }
+    return inside;
+  }
+
+  /**
+   * Throws exception for invalid coordinates.
+   */
+  static List<ParseGeoPoint> _validate(List<ParseGeoPoint> coords) {
```

Comment:
I was thinking about using this for offline query in the future and wanted it to be exposed. I can still remove the _.